# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/saba-naseem22/my-ml-repo/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

Unit of Analysis: one row = one unique content piece / search query performance for a client.  
Time Window: Mid-panel month `2026-03` (March 1 to March 31, 2026)

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. Fields: feature / label / context / excluded
features: `impressions`,`clicks`,`position`,`word_count`,`ctr`
label:`refresh_needed`(1=high impression-low clicks,0=healty)
context: `client_id`,`content_id`
excluded:future performance metrics post march 2026(to avoid temporary data leakage).




In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [3]:
import pandas as pd
from google.colab import userdata
from huggingface_hub import hf_hub_download

token = userdata.get("HF_TOKEN")

file_path = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    filename="fact_content_daily_performance/month=2026-03/data_0.parquet",
    repo_type="dataset",
    token=token
)

df = pd.read_parquet(file_path)

print(df.head())
print(df.columns)
print(df.shape)

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

  report_date           client_hash_id           content_hash_id  \
0  2026-03-01  client_73cda7b4e4f265ea  content_b7e512995f79d5a6   
1  2026-03-01  client_73cda7b4e4f265ea  content_05597932fe4da067   
2  2026-03-01  client_73cda7b4e4f265ea  content_7a105f548d9c6916   
3  2026-03-01  client_73cda7b4e4f265ea  content_905aa32a0230694e   
4  2026-03-01  client_73cda7b4e4f265ea  content_a3ea9792f793ec72   

   client_has_gsc  client_has_ga4  gsc_data_available ga4_data_available  \
0            True           False                True               None   
1            True           False                True               None   
2            True           False                True               None   
3            True           False                True               None   
4            True           False                True               None   

   gsc_impressions  gsc_clicks  gsc_sum_position  ...  sessions_paid  \
0               20           0                67  ...         

In [4]:
print(df.duplicated(
    subset=["report_date","client_hash_id","content_hash_id"]
).sum())

0


In [5]:
print("Rows:", len(df))
print("Start:", df["report_date"].min())
print("End:", df["report_date"].max())

Rows: 9841378
Start: 2026-03-01
End: 2026-03-31


In [6]:
active = df[df["gsc_data_available"] == True]

print("Rows with GSC data available:", len(active))

Rows with GSC data available: 3611061


## 4. Data limits
`balanced history`: dataset covers onlysingle month,so long term trends cananot be evaluated
`partial integration`:out of 9841378 total rows 3611061 have GSC available.
`missing GA4 data`:early rows  lack full integration, means decision support queries must account for NULl values.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.